File: solution/java/week06/exercise2/src/main/java/report/ReportGenerator.java
package report;

import io.vavr.control.Option;
import io.vavr.control.Try;

import java.util.List;
import java.util.Map;

class ReportGenerator {
    private final Map<ReportType, CanGenerateReport> reportGenerators;

    public ReportGenerator() {
        reportGenerators = Map.of(
                ReportType.CSV, new CsvReportGenerator(),
                ReportType.PDF, new PdfReportGenerator()
        );
    }

    public Try<String> generateReport(ReportType reportType, List<ReportData> data) {
        if (data == null || data.isEmpty()) {
            return Try.failure(new NoDataToReportException("No data provided for report generation."));
        }

        return Option.of(reportGenerators.get(reportType))
                .map(generator -> Try.of(() -> generator.generateReport(data)))
                .getOrElse(Try.failure(new UnsupportedReportTypeException("Report type " + reportType + " not supported.")));
    }

}

File: solution/ts/week06-legacy-code/exercise2/src/reportGenerator.ts
import {fold, fromNullable} from 'fp-ts/lib/Option';

import {NoDataToReportError} from './noDataToReportError';
import {Either, left, right} from "fp-ts/Either";
import {ReportType} from "./reportType";
import {CanGenerateReport} from "./canGenerateReport";
import {ReportData} from "./reportData";
import {UnsupportedReportTypeError} from "./unsupportedReportTypeError";

export class ReportGenerator {
    constructor(private reportGenerators: Map<ReportType, CanGenerateReport>) {
    }

    generateReport(reportType: ReportType, data: ReportData[]): Either<Error, string> {
        if (!data || data.length === 0) {
            return left(new NoDataToReportError("No data provided for report generation."));
        }

        return fold(
            () => left(new UnsupportedReportTypeError("Report type " + reportType + " not supported.")),
            (g: CanGenerateReport) => right(g.generateReport(data))
        )(fromNullable(this.reportGenerators.get(reportType)));
    }
}